File: packages/manifest/src/state.ts
import type { AgentManifest, AgentState, InputSchema, OutputMapping, StepRef } from "./types.js";
import { resolvePath, interpolate } from "./template.js";

/**
 * Normalize an agent ID into a valid state key.
 * "my-long-agent-name" → "myLongAgentName"
 */
export function normalizeId(id: string): string {
  return id.replace(/-([a-z])/g, (_, c) => c.toUpperCase());
}

/**
 * Get the state key for a step: explicit stateKey on the step,
 * or explicit stateKey on the inline agent, or normalized agent id/ref.
 */
export function getStateKey(step: StepRef): string {
  if (step.stateKey) return step.stateKey;
  if (step.agent?.stateKey) return step.agent.stateKey;
  if (step.ref) return normalizeId(step.ref);
  if (step.agent) return normalizeId(step.agent.id);
  return "unknown";
}

/**
 * Returns true if this step is a reference (not inline).
 */
export function isRefStep(step: StepRef): step is StepRef & { ref: string } {
  return typeof step.ref === "string";
}

/**
 * Create initial state from input and an inputSchema.
 * If no inputSchema, wraps the raw input as { userQuery: input }.
 */
export function createInitialState(input: unknown, inputSchema?: InputSchema): AgentState {
  if (!inputSchema) {
    // Default: plain string input
    return { userQuery: typeof input === "string" ? input : String(input) };
  }

  // Structured input: input should be an object
  if (typeof input === "object" && input !== null) {
    const state: AgentState = {};
    for (const [key, def] of Object.entries(inputSchema)) {
      const provided = (input as Record<string, unknown>)[key];
      if (provided !== undefined) {
        state[key] = provided;
      } else {
        // Apply default if defined, otherwise null
        const defaultValue = typeof def === "object" ? def.default : undefined;
        state[key] = defaultValue ?? null;
      }
    }
    return state;
  }

  // inputSchema declared but input is a string — wrap as userQuery fallback
  return { userQuery: String(input) };
}

/**
 * Apply a step's input transform: maps parent state to child agent's input.
 * If no transform is provided, the child's input is the upstream value
 * (parent's input for the first step, the previous step's output otherwise).
 */
export function applyInputTransform(
  transform: Record<string, string> | undefined,
  parentState: AgentState,
  defaultUpstream: unknown
): unknown {
  if (!transform) return defaultUpstream;

  const result: Record<string, unknown> = {};
  for (const [key, template] of Object.entries(transform)) {
    // If the template is a simple {{ref}}, resolve the value directly (preserving type)
    const simpleMatch = template.match(/^\{\{(.+?)\}\}$/);
    if (simpleMatch) {
      result[key] = resolvePath(parentState, simpleMatch[1].trim()) ?? null;
    } else {
      // Complex template with mixed text + refs: interpolate as string
      result[key] = interpolate(template, parentState);
    }
  }
  return result;
}

/**
 * Apply an output mapping: maps state values to output properties.
 * Supports nested output objects.
 */
export function applyOutputMapping(
  mapping: OutputMapping,
  state: AgentState
): Record<string, unknown> {
  const result: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(mapping)) {
    if (typeof value === "string") {
      // Template reference
      const simpleMatch = value.match(/^\{\{(.+?)\}\}$/);
      if (simpleMatch) {
        result[key] = resolvePath(state, simpleMatch[1].trim()) ?? null;
      } else {
        result[key] = interpolate(value, state);
      }
    } else {
      // Nested object mapping
      result[key] = applyOutputMapping(value, state);
    }
  }
  return result;
}

File: python/src/agntz/manifest/state.py
"""State helpers for manifest pipeline execution."""

from __future__ import annotations

import re
from typing import Any

from .template import interpolate, resolve_path
from .types import AgentManifest, AgentManifestBase, AgentState, InputSchema, OutputMapping, StepRef

_SIMPLE_TEMPLATE_RE = re.compile(r"^\{\{(.+?)\}\}$")


def normalize_id(agent_id: str) -> str:
    parts = [part for part in agent_id.split("-") if part]
    if not parts:
        return agent_id
    return parts[0] + "".join(part[:1].upper() + part[1:] for part in parts[1:])


def get_manifest_state_key(manifest: AgentManifest) -> str:
    return manifest.state_key or normalize_id(manifest.id)


def get_state_key(step: StepRef) -> str:
    if step.state_key:
        return step.state_key
    agent = step.agent
    if isinstance(agent, dict):
        state_key = agent.get("stateKey") or agent.get("state_key")
        if isinstance(state_key, str):
            return state_key
        agent_id = agent.get("id")
        if isinstance(agent_id, str):
            return normalize_id(agent_id)
    if isinstance(agent, AgentManifestBase):
        if agent.state_key:
            return agent.state_key
        return normalize_id(agent.id)
    if step.ref:
        return normalize_id(step.ref)
    return "unknown"


def create_initial_state(input_value: Any, input_schema: InputSchema | None = None) -> AgentState:
    if not input_schema:
        return {"userQuery": input_value if isinstance(input_value, str) else str(input_value)}

    if isinstance(input_value, dict):
        state: AgentState = {}
        for key, definition in input_schema.items():
            if key in input_value:
                state[key] = input_value[key]
            elif isinstance(definition, dict) and "default" in definition:
                state[key] = definition["default"]
            else:
                state[key] = None
        return state

    return {"userQuery": str(input_value)}


def apply_input_transform(
    transform: dict[str, str] | None,
    parent_state: AgentState,
    default_upstream: Any,
) -> Any:
    if not transform:
        return default_upstream

    result: dict[str, Any] = {}
    for key, template in transform.items():
        match = _SIMPLE_TEMPLATE_RE.match(template)
        if match:
            result[key] = resolve_path(parent_state, match.group(1).strip())
        else:
            result[key] = interpolate(template, parent_state)
    return result


def apply_output_mapping(mapping: OutputMapping, state: AgentState) -> dict[str, Any]:
    result: dict[str, Any] = {}
    for key, value in mapping.items():
        if isinstance(value, str):
            match = _SIMPLE_TEMPLATE_RE.match(value)
            result[key] = (
                resolve_path(state, match.group(1).strip())
                if match
                else interpolate(value, state)
            )
        elif isinstance(value, dict):
            result[key] = apply_output_mapping(value, state)
        else:
            result[key] = value
    return result
